### PR TITLE
firestore: query watch samples

### PR DESCRIPTION
**Pull request by AlisskaPie**
Samples:
- [x] Listen to a document 
- [x] Listen to multiple documents in a collection
- [x] View changes between snapshots
- [x] Detach a listener
- [x] Handle listen errors

### CODE REVIEW

File: firestore/firestore_snippets/listen_changes.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"cloud.google.com/go/firestore"
+)
+
+// listenChanges listens to a query, returning the list of DocumentChange
+// in the first snapshot.
```
Comment (IlyaFaer) [inline firestore/firestore_snippets/listen_changes.go]: What do you mean, the first snapshot? It's the snapshot with new changes, so it's not the first of something in my vision

File: firestore/firestore_snippets/listen_test.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	}
+
+	ctx := context.Background()
+	client, err := firestore.NewClient(ctx, projectID)
+	if err != nil {
+		t.Fatalf("firestore.NewClient: %v", err)
+	}
+	defer client.Close()
+
+	// Set prepareQuery for all listen tests.
```
Comment (IlyaFaer) [inline firestore/firestore_snippets/listen_test.go]: Not sure what does this comment mean

File: firestore/firestore_snippets/listen_test.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	if projectID == "" {
+		t.Skip("Skipping firestore test. Set GOLANG_SAMPLES_FIRESTORE_PROJECT.")
+	}
+
+	ctx := context.Background()
+	client, err := firestore.NewClient(ctx, projectID)
+	if err != nil {
+		t.Fatalf("firestore.NewClient: %v", err)
+	}
+	defer client.Close()
```
Comment (IlyaFaer) [inline firestore/firestore_snippets/listen_test.go]: Project and client creation could be moved into separate function

File: firestore/firestore_snippets/listen_changes.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+	// projectID := "project-id"
+	ctx := context.Background()
+
+	client, err := firestore.NewClient(ctx, projectID)
+	if err != nil {
+		return fmt.Errorf("firestore.NewClient: %v", err)
+	}
+	defer client.Close()
+
+	snap, err := client.Collection("cities").Where("state", "==", "CA").Snapshots(ctx).Next()
```
Comment (IlyaFaer) [inline firestore/firestore_snippets/listen_changes.go]: `Next()` probably should be called within `for`. For now it'll only listen for a single one snapshot, but I suspect the samples intended to show the endless listening

File: firestore/firestore_snippets/listen_document.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+	for {
+		snap, err := dsnap.Next()
+		if !snap.Exists() {
+			return fmt.Errorf("current data: null")
+		}
+		if err != nil {
+			return fmt.Errorf("listen failed: %v", err)
+		}
+		fmt.Fprintf(w, "Received document snapshot: %v\n", snap.Data())
+		return nil
```
Comment (IlyaFaer) [inline firestore/firestore_snippets/listen_document.go]: `Next()` seems to be blocking, so for single snapshot we can just call `Next()`, without `for` cycle, like in `listen_errors`

File: firestore/firestore_snippets/listen_changes.go (hunk 1 above)
Comment (AlisskaPie) [reply firestore/firestore_snippets/listen_changes.go]: When we're iterating through `iter.Next()` just once, we get a first snapshot.
Also, I copied this comment from Java samples, they're similar.
[listenForChanges](https://github.com/GoogleCloudPlatform/java-docs-samples/blob/fe0d5cb7214680461928cac1d93d1711842884b9/firestore/src/main/java/com/example/firestore/snippets/ListenDataSnippets.java#L124)

File: firestore/firestore_snippets/listen_changes.go (hunk 4 above)
Comment (AlisskaPie) [reply firestore/firestore_snippets/listen_changes.go]: I watched the same samples on other languages, and it looks like they use the same way.
I tried to listen with for loop but, as I wrote you personally, I got freezes and falls for local test.

File: firestore/firestore_snippets/listen_test.go (hunk 2 above)
Comment (AlisskaPie) [reply firestore/firestore_snippets/listen_test.go]: Once we run this test, we'll have data in base, that can be used for other tests.
I saw something similar in Storage samples. There we need to run create bucket test first before run other tests.
I could remove this comment, I was not sure about adding it

File: firestore/firestore_snippets/listen_test.go (hunk 3 above)
Comment (AlisskaPie) [reply firestore/firestore_snippets/listen_test.go:77]: I thought about it, but I don't have idea how to skip test in a separate function, so I left it as it's.
Do you have an idea?

File: firestore/firestore_snippets/listen_errors.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"cloud.google.com/go/firestore"
+)
+
+// listenErrors demonstrates how to handle listening errors.
+func listenErrors(w io.Writer, projectID string) error {
```
Comment (AlisskaPie) [inline firestore/firestore_snippets/listen_errors.go]: I'm not sure about this test, but I did the same as was done in Java

File: firestore/firestore_snippets/listen_document.go (hunk 5 above)
Comment (AlisskaPie) [reply firestore/firestore_snippets/listen_document.go]: What do you mean by `Next()` is blocking?
Ok, I'll fix it.

File: firestore/firestore_snippets/listen_changes.go (hunk 1 above)
Comment (AlisskaPie) [reply firestore/firestore_snippets/listen_changes.go]: How do you suggest to rewrite this comment?

File: firestore/firestore_snippets/listen_document.go (hunk 5 above)
Comment (AlisskaPie) [reply firestore/firestore_snippets/listen_document.go]: Done

File: firestore/firestore_snippets/listen_test.go (hunk 2 above)
Comment (AlisskaPie) [reply firestore/firestore_snippets/listen_test.go]: Removed

File: firestore/firestore_snippets/listen_test.go (hunk 3 above)
Comment (AlisskaPie) [reply firestore/firestore_snippets/listen_test.go]: Done

File: firestore/firestore_snippets/listen_changes.go (hunk 1 above)
Comment (IlyaFaer) [reply firestore/firestore_snippets/listen_changes.go]: If it's taken from Java then let's keep it

File: firestore/firestore_snippets/listen_changes.go (hunk 4 above)
Comment (IlyaFaer) [reply firestore/firestore_snippets/listen_changes.go]: Looking at Python sample I'd say it registers a callback function, so on every new snapshot it'll call the callback func (cycle). In Go there is no registering, but moving the `Next()` call into `for` cycle will do the same thing. Seems to me all the `Next()` calls must be moved into `for` cycles, or it'll be listening just for a single one change

What are the local tests failures?

File: firestore/firestore_snippets/listen_changes.go (hunk 1 above)
Comment (AlisskaPie) [reply firestore/firestore_snippets/listen_changes.go]: In Java they have: Listen to a query, returning the list of DocumentChange **events** in the first snapshot.
They have event methods and we don't so I decided to delete **events**.

**Pull request closed without merging**